cluster/v1alpha1/types.go

```diff
@@ -157,3 +160,230 @@ var ReservedClusterClaimNames = [...]string{
+	//    with expected number of ManagedClusters will be selected;
+	// 3) If the nubmer of matched ManagedClusters is equal to NumberOfClusters, all matched ManagedClusters
+	//    will be selected;
+	// 4) If the nubmer of matched ManagedClusters is less than NumberOfClusters, all matched ManagedClusters
+	//    will be selected, and the status of condition `PlacementConditionSatisfied` will be set to false;
+	// +optional
+	NumberOfClusters *int32 `json:"numberOfClusters,omitempty"`
+
+	// RequiredDuringSchedulingRequiredDuringExecution represents a scheduling strategy for a
+	// cluster selector. If specified, the cluster selector should be matched at both scheduling
+	// time (when the placement decisions are made for the first time) and execution time (After
+	// one or multiple placement decisions are made). It means,
+	// 1) Any ManagedCluster, which does not match the selector, should not be selected;
+	// 2) If a selected ManagedCluster ceases to match the selector (e.g. due to an update) after the
+	//    placement decisions are made, it will be eventually removed from the placement decisions;
+	// 3) If a ManagedCluster starts to match the selector after the placement decisions are made, it
+	//    will either be selected or at least has chance to be selected (when NumberOfClusters is
+	//    specified);
+	//
+	// Other scheduling strategies not yet implemented includes
+	// RequiredDuringSchedulingIgnoredDuringExecution and PreferredDuringSchedulingIgnoredDuringExecution
+	// +optional
+	RequiredDuringSchedulingRequiredDuringExecution *ClusterSelector `json:"requiredDuringSchedulingRequiredDuringExecution,omitempty"`
+}
+
+// ClusterSelector represents the AND of the containing selectors
+type ClusterSelector struct {
+	// LabelSelector represents a selector of ManagedClusters by label
+	// +optional
+	LabelSelector metav1.LabelSelector `json:"labelSelector,omitempty"`
+
+	// ClaimSelector represents a selector of ManagedClusters by clusterClaims in status
+	// +optional
+	ClaimSelector ClusterClaimSelector `json:"claimSelector,omitempty"`
+}
+
+// ClusterClaimSelector is a claim query over a set of ManagedClusters. An empty cluster claim
+// selector matches all objects. A null cluster claim selector matches no objects.
+type ClusterClaimSelector struct {
+	// matchExpressions is a list of cluster claim selector requirements. The requirements are ANDed.
+	// +optional
+	MatchExpressions []metav1.LabelSelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// Affinity is a group of affinity scheduling rules.
+type Affinity struct {
+	// ClusterAntiAffinity describes cluster affinity scheduling rules for the placement.
+	// +optional
+	ClusterAntiAffinity *ClusterAntiAffinity `json:"clusterAntiAffinity,omitempty"`
+}
+
+// ClusterAntiAffinity is a group of cluster affinity scheduling rules.
+type ClusterAntiAffinity struct {
```

Comment:
Recommend getting someone associated with kube-scheduler or node to consider this API carefully.  Perhaps @damemi (scheduler) or @sjenning (node)

Reply:
this seems like a really similar goal to [descheduler](https://github.com/kubernetes-sigs/descheduler), which was also considered as a place to implement `Required/Required`. I don't think `RequiredDuringSchedulingRequiredDuringExecution` is implemented anywhere else in the upstream scheduler, and I'm not familiar with this project, so is your goal to essentially add that functionality here?

Reply:
@damemi Yes, we are going to implement Required/Required in this API. I know it may introduced some potential issues, like
- Bad performance. Re-calculation is required once managed clusters change.
- Placement decisions changes frequently

Do you have any other concerns on implementing such a feature?

Reply:
No, I just didn't know if you were aware of the descheduler and curious if you had any other similar goals. 

Because this isn't implemented as a pod spec option upstream (https://github.com/kubernetes/api/blob/3a2d6b5bb7f84daa2125131865eb81b151448d5b/core/v1/types.go#L2746) this will be entirely on your end. That could expose you to some problems if this is later implemented upstream, and it conflicts with your implementation. But I think that risk is low